src/Common/src/System/CommonRuntimeTypes.cs

```diff
@@ -31,8 +31,8 @@ internal static class CommonRuntimeTypes
         internal static Type IntPtr { get { return s_intPtr; } }
         internal static Type Single { get { return s_single; } }
         internal static Type Double { get { return s_double; } }
-        internal static Type Decimal { get { return s_decimal; } }
-        internal static Type DateTime { get { return s_datetime; } }
+        internal static Type Decimal { get { return NotSoCommonTypes.s_decimal; } }
```

Comment:
It would be better to deleted these and convert the few places that use it to use `typeof`, e.g. `if (type == typeof(Decimal))`.

Caching these in statics is actually a deoptimization.

The JIT has optimization for `operator==` on System.Type that kicks in when one of the sides is typeof. It should compile this into a simple vtable compare.

Reply:
It would be nice to port the fix for this deoptimization to CoreCLR as well.

Reply:
I take it, the comment at the top of this class only applies for .NET Native?

Reply:
I actually have a hard time finding the places where RyuJIT would be able to optimize the type comparison - most of these are used in deep generic code where we have a `System.Type` from who knows where that we want to compare to a well known one, or we want to grab a specific `System.Type` to return.

I think typeof is still more expensive on CoreRT than on CoreCLR. It's at minimum a hashtable lookup. CoreCLR can do it faster for some reason.

Reply:
`CommonTypes.Decimal` is used in a single place in https://github.com/dotnet/corert/blob/09682e3a157ef6b52d0c20401d66ee650848aca4/src/System.Private.CoreLib/src/Internal/Reflection/Augments/ReflectionAugments.cs#L90

If you change this `if (type == typeof(Decimal))`, the RyuJIT should optimize it to a simple `cmp` instruction (look for `CORINFO_INTRINSIC_TypeEQ`).

I am not suggestion to get rid of this class completely. I agree that it is not straightforward to get rid of this completely. I am just suggestion to get rid of the ones you are moving to `NotSoCommonTypes`.

Reply:
The equivalent in CoreCLR that can use the same fix is here: https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/RuntimeType.cs#L192

Reply:
Hmmm, the optimization is not as good as I though. It is not able to get rid of the materializing of the Type object.

Reply:
> Hmmm, the optimization is not as good as I though. It is not able to get rid of the materializing of the Type object.

We could probably come up with a scheme for RyuJIT to be able optimize this.

`typeof(X)` expands to a set of calls "load native `TypeHandle` → get `RuntimeTypeHandle` from native `TypeHandle` -> `Type.GetTypeFromHandle`. RyuJIT optimizes this entire sequence into a single `CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE` helper call that takes the native `TypeHandle` as an argument.

RyuJIT can then optimize this to a simple compare if the other side of the comparison is a another `typeof` or `someObject.GetType()`. But that's not the case here.

If we could add a helper to compare native `TypeHandle` with an arbitrary `System.Type`, we could have RyuJIT call that in that case. That would help avoid materializing the `System.Type` from the native `TypeHandle`.

I could try to prototype that.

Reply:
dotnet/runtime#31686 is the JIT fix to optimize the pattern and make a lot of `CommonRuntimeTypes` obsolete.